crates/ruff/src/rules/pylint/rules/redefined_loop_name.rs

```diff
@@ -0,0 +1,325 @@
+        )
+    }
+}
+
+struct ExprWithBindingKind<'a> {
+    expr: &'a Expr,
+    binding_kind: BindingKind,
+}
+
+struct InnerForWithAssignTargetsVisitor<'a> {
+    dummy_variable_rgx: &'a HashableRegex,
+    assignment_targets: Vec<ExprWithBindingKind<'a>>,
+}
+
+impl<'a, 'b> Visitor<'b> for InnerForWithAssignTargetsVisitor<'a>
+where
+    'b: 'a,
+{
+    fn visit_stmt(&mut self, stmt: &'b Stmt) {
+        // Collect target expressions.
+        match &stmt.node {
+            // For and async for.
+            StmtKind::For { target, .. } | StmtKind::AsyncFor { target, .. } => {
+                self.assignment_targets.extend(
+                    assignment_targets_from_expr(target, self.dummy_variable_rgx).map(|expr| {
+                        ExprWithBindingKind {
+                            expr,
+                            binding_kind: BindingKind::For,
+                        }
+                    }),
+                );
+            }
+            // With.
+            StmtKind::With { items, .. } => {
+                self.assignment_targets.extend(
+                    assignment_targets_from_with_items(items, self.dummy_variable_rgx).map(
+                        |expr| ExprWithBindingKind {
+                            expr,
+                            binding_kind: BindingKind::With,
+                        },
+                    ),
+                );
+            }
+            // Assignment, augmented assignment, and annotated assignment.
+            StmtKind::Assign { targets, .. } => {
+                self.assignment_targets.extend(
+                    assignment_targets_from_assign_targets(targets, self.dummy_variable_rgx).map(
+                        |expr| ExprWithBindingKind {
+                            expr,
+                            binding_kind: BindingKind::Assignment,
+                        },
+                    ),
+                );
+            }
+            StmtKind::AugAssign { target, .. } | StmtKind::AnnAssign { target, .. } => {
+                self.assignment_targets.extend(
+                    assignment_targets_from_expr(target, self.dummy_variable_rgx).map(|expr| {
+                        ExprWithBindingKind {
+                            expr,
+                            binding_kind: BindingKind::Assignment,
+                        }
+                    }),
+                );
+            }
+            _ => {}
+        }
+        // Decide whether to recurse.
+        match &stmt.node {
+            // Don't recurse into blocks that create a new scope.
+            StmtKind::ClassDef { .. } => {}
+            StmtKind::FunctionDef { .. } => {}
+            // Otherwise, do recurse.
+            _ => {
+                visitor::walk_stmt(self, stmt);
+            }
+        }
+    }
+}
+
+fn assignment_targets_from_expr<'a, U>(
+    expr: &'a Expr<U>,
+    dummy_variable_rgx: &'a HashableRegex,
+) -> Box<dyn Iterator<Item = &'a Expr<U>> + 'a> {
```

Comment:
Another approach would be to pass in a mutable vector, and have the various arms push to it. Would that work? Perhaps not as nice conceptually, but a bit simpler.

Reply:
Certainly could, but may be less efficient overall because it will require an additional temporary vector for each call. This is because the outputs of these methods can't be pushed directly to the InnerForWithAssignTargetsVisitor.assignment_targets vector as they first need to be paired with the appropriate BindingKind variant by the caller, which depends on the call path; so the intermediate Vec<Expr<U>> would be immediately discarded each time.

On the other hand, I suppose the Boxes also necessitate a heap allocation, albeit one that has a fixed size and doesn't get expanded...

An alternative would be to pass in both a &mut Vec, and also the BindingKind, to these methods, and have them push ExprWithBindingKind into the vector.